File: C++/Maximize The Cut Segments.cpp
class Solution
{
    public:
    int maximizeTheCuts(int n, int x, int y, int z)
    {
        vector<int> dp(n+1, 0);
        for(int i = n; i >= 0; i--){
            if(i != n && !dp[i]) continue;
            if(i>=x) dp[i-x] = max(dp[i-x], 1 + dp[i]);
            if(i>=y) dp[i-y] = max(dp[i-y], 1 + dp[i]);
            if(i>=z) dp[i-z] = max(dp[i-z], 1 + dp[i]);
        }
        return dp[0];
    }
};

File: JAVA/Maximize The Cut Segments.java
class Solution
{
      public int maximizeCuts(int n, int x, int y, int z)
    {
       int[] dp = new int[n + 1];
        Arrays.fill(dp, Integer.MIN_VALUE);
        dp[0] = 0;
        for (int i = 1; i <= n; i++) {
            if (i >= x && dp[i - x] != Integer.MIN_VALUE) {
                dp[i] = Math.max(dp[i], dp[i - x] + 1);
            }
            if (i >= y && dp[i - y] != Integer.MIN_VALUE) {
                dp[i] = Math.max(dp[i], dp[i - y] + 1);
            }
            if (i >= z && dp[i - z] != Integer.MIN_VALUE) {
                dp[i] = Math.max(dp[i], dp[i - z] + 1);
            }
        }
        return dp[n] < 0 ? 0 : dp[n];
    }
}
